agent/sshd/utmp.go

```diff
@@ -0,0 +1,273 @@
+package sshd
+
+/*	The Utmpx struct is derived from the Linux definition (man 5 utmpx).
+
+	At session start, a utmp record is constructed containing the PID,
+	tty name, ID, username, remote host, source IP and time.
+	Type is set to UserProcess and the record is written to UtmpxFile.
+	If a record with the same ID exists, that record is overwritten;
+	otherwise the record is appended to the file.  The same record is
+	appended to WtmpxFile.
+
+	At session end, Type is set to DeadProcess, the User and Host fields
+	are cleared, time is updated and the record is wrtten to UtmpxFile,
+	overwriting the session start record. The same record with ID and
+	source IP address cleared is appended to WtmpxFile.
+*/
+
+import (
+	"os"
+	"unsafe"
+	"bytes"
+	"net"
+	"encoding/binary"
+	"golang.org/x/sys/unix"
+	"github.com/sirupsen/logrus"
+)
+
+type ExitStatus struct {
+	ETermination	int16		// Process temination status - not used
+	EExit		int16		// Process exit status - not used
+}
+
+type Utmpx struct {
+	Type		int16		// UserProcess or DeadProcess
+	Padding		[2]byte		// Padding to align rest of struct
+	Pid		int32		// PID of the ShellHub agent
+	Line		[32]byte	// tty associated with the process
+	ID		[4]byte		// Index, last 4 characters of Line
+	User		[32]byte	// Username
+	Host		[256]byte	// Source IP address
+	Exit		ExitStatus	// Exit status - not used
+	Session		int32		// Session ID - not used
+	Tv		unix.Timeval	// Time entry was made
+	AddrV6		[4]uint32	// Source IP address. IPv4 in AddrV6[0]
+	Reserved	[20]byte	// Not used
+}
+
+const (
+	UtmpxFile   = "/var/run/utmp"
+	WtmpxFile   = "/var/log/wtmp"
+	UserProcess = 0x7		// Normal process
+	DeadProcess = 0x8		// Terminated process
+)
+
+// This function updates the utmp and wtmp files at the start of a user session
+func utmpStartSession(line, user, remoteAddr string) Utmpx {
+
+	var u Utmpx
+
+	u.Type = UserProcess
+	a := unix.Timeval{}
+	unix.Gettimeofday(&a)
+	u.Tv.Sec, u.Tv.Usec = a.Sec, a.Usec
+	u.Pid = int32(os.Getpid())
+
+	// remoteAddr has the form <IPv4 address>:<port>
+	// or [<IPv6 address>]:port
+
+	// Remove the port suffix and also the square brackets
+	// if IPv6, leaving the bare IPv4 or IPv6 address
+
```

Comment:
Drop the empty line and make this a single comment block such as:

```go
// foo
//
// bar
```

Reply:
Done.